src/vaults/VaultV2.sol

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.16;
+
+import {MathUpgradeable} from "@openzeppelin/contracts-upgradeable/utils/math/MathUpgradeable.sol";
+import {ERC20} from "solmate/src/tokens/ERC20.sol";
+import {SafeTransferLib} from "solmate/src/utils/SafeTransferLib.sol";
+
+import {Vault} from "src/vaults/Vault.sol";
+import {NftGate} from "src/vaults/NftGate.sol";
+import {HarvestStorage} from "src/vaults/HarvestStorage.sol";
+import {BaseStrategy as Strategy} from "src/strategies/BaseStrategy.sol";
+import {uncheckedInc} from "src/libs/Unchecked.sol";
+import {VaultErrors} from "src/libs/VaultErrors.sol";
+
+contract VaultV2 is Vault, NftGate, HarvestStorage {
+    using SafeTransferLib for ERC20;
+    using MathUpgradeable for uint256;
+
+    function _deposit(address caller, address receiver, uint256 assets, uint256 shares) internal virtual override {
+        _checkNft(receiver);
+        if (shares == 0) revert VaultErrors.ZeroShares();
+        _mint(receiver, shares);
+        _asset.safeTransferFrom(caller, address(this), assets);
+        emit Deposit(caller, receiver, assets, shares);
+    }
+
+    function _getWithdrawalFee(uint256 assets, address owner) internal view virtual override returns (uint256) {
+        uint256 feeBps;
+        if (nftDiscountActive && accessNft.balanceOf(owner) > 0) {
+            feeBps = withdrawalFeeWithNft;
+        } else {
+            feeBps = withdrawalFee;
+        }
+        return assets.mulDiv(feeBps, MAX_BPS, MathUpgradeable.Rounding.Up);
+    }
+
+    function harvest(Strategy[] calldata strategyList) external virtual override onlyRole(HARVESTER) {
```

Comment:
Vault already has harvest, can we minimize overlapping?

Reply:
We have to override the function in order to change the implementation.